design/queue.md

```diff
@@ -0,0 +1,251 @@
+# Executor Queue
+
+## Context
+
+This is an executor plugin for Screwdriver that makes use of [Resque][node-resque-URL] to add a queueing mechanism.
+
+#### Problem
+
+Right now, Screwdriver directly calls the executor when a build needs to be started. If the executor target is offline or full, the build is generally dropped with an error message. This isn't a good user experience, especially if the error is hidden due to the build being started from a PR or commit.
+
+Instead, Screwdriver should be pushing planned builds into a queue; all available executors should be listening on that queue and taking items off when they have capacity. This will allow us to distribute between N number of Docker Swarm clusters, for example.
+
+The rest of this document will describe the overall architecture of the executor queue, as well as of the various plug-ins that will eventually be built on top of it.
+
+## Design Decisions
+
+### Requirements
+
+* Appropriate license (Apache 2.0)
+* Includes plugins to prevent duplicates from entering the queue
+* Handles large queue size
+* Allows for thousands of queries per seconds (queues and dequeues quickly)
+* Support for logging, stats, and metrics
+* Distributed load
+* Persistent
+* Allows for prioritization
+* Password-protected security
+
+### Why Resque?
+
+We evaluated numerous queueing mechanisms based on the acceptance criteria above. We eventually settled on [Node-resque](node-resque-URL) because it meets all the requirements above and fits in well with the rest of our codebase. In addition, the community around it is pretty active, and we should easily be able to implement a variety of plugins for the queue.
+
+#### Resque Plugins
+
+A plugin can have access to 4 unique hooks:
```

Comment:
Based on the [Resque documentation](https://github.com/resque/resque/blob/master/docs/HOOKS.md#job-hooks), it has access to more than 4